### PR TITLE
New Style language, system runtime, and shape definitions

### PR DESCRIPTION
**Pull request by hypotext**
# Description

Related issues: #97, #99 (PR for old `new-style` branch containing debugging commits)

Implement a new design of the Style language, which has language features such as a new selector grammar and redesigned Style semantics.

# Implementation strategy and design decisions

See the Style section of the design paper and this [wiki page](https://github.com/penrose/penrose/wiki/Testing).

# Examples with steps to reproduce them

See the examples [wiki page](https://github.com/penrose/penrose/wiki/Examples) as well as the [LA domain page](https://github.com/penrose/penrose/wiki/The-LA-Domain-Full-Spec).

# Not implemented in this PR

- [ ] Derived properties
- [ ] Layering
- [ ] Matching WRT subtyping
- [ ] Raytracing semantics
- [ ] Lists and list accesses; tuples
- [ ] Matching anon exprs (make anonymous variables for unification for Substance program)
- [ ] ShapeDef / Function typechecking
- [ ] Block statics
- [ ] Selector specificity
- [ ] Tests for the new features (beyond the wiki examples)
 
# Past TODOs (7/17)

**bolded** = highest priority

**Port**
- [x] **Shape definitions**
  - [x] Circle
  - [x] Arrow
  - [x] Curve (CubicBezier?)
  - [x] Line
  - [x] Square
  - [x] Arc
  - [x] Rectangle
  - [x] Parallelogram
  - [x] Text
  - [x] Image
  - [x] Ellipse
  - [x] Point (AnchorPoint?)
  - [x] Their samplers
- [ ] Shapes' derived properties
- [x] **Shape rendering code**
  - [x] Circle
  - [x] Arrow
  - [x] Curve (CubicBezier?)
  - [x] Line
  - [x] Square
  - [x] Rectangle
  - [x] Parallelogram (needs @dormaayan to clean up code first)
  - [x] Text
  - [x] Image
  - [x] Ellipse
  - [x] AngleMark  (needs @dormaayan to clean up code first)
  - [x] Point (AnchorPoint?)
- [x] **Objective functions** 
- [x] **Constraints**
- [x] **Computations**
- [x] **Test**
  - [x] Shapes
  - [x] Objectives
  - [x] Constraints
  - [x] Computations

Compiler + front-end
- [x] **Labels and text**
- [x] Pass frontend info back to optimizer

**Style compiler (1)**
- [x] Namespaces (+ global)
- [x] Generate names for anonymous fields (don't belong to a specific Sub object). Does the parser need to change?
  - [x] Fix programs that use cartesian selectors (e.g. venn.sty, `Set x; Set y`) 
- [ ] Derived properties
  - [ ] Design + implement + test feature, then port Style programs

**Style compiler (2)**
- [ ] Layering
- [x] Matching WRT equality and implication
  - [x] Add/use beta environment
  - [x] Test matching more thoroughly
- [ ] Match WRT subtyping

**Style compiler (3)**
- [ ] **Raytracing semantics**
- [ ] Implement lists and accesses, as well as tuples (Pt2)
- [ ] Matching anon exprs (make anonymous variables for unification for Substance program)

**Style compiler (misc)**
- [x] Figure out if we need to check substitution validity
- [x] Document what `T x; T y` should match (all 4 combinations or not?)
- [ ] (*ShapeDef* related) use `invoke*` functions defined in `ShapeDef` to type check the objectives/constraints/comps. 

Optimization
- [x] **Get "contains" working**
- [x] Sizefuncs on everything, or canvas bbox contains everything

Shapes
- [ ] Derived properties
- [ ] Parse `.shapes` file
- [ ] Splice in rendering code
- [ ] Composite shapes

End-to-end
- [x] **Repro wiki examples**
- [x] **Get LA example working**
- [ ] **Get RA example working**

Documentation/cleanup
- [x] **Document the three representations (translation, varying state, shapes) and information flow between them**
- [x] Document sampling/setting logic
- [x] Improve error and warning output (i.e. so they aren't buried in the translation)
- [x] Clean up/standardize error messages and warnings
- [x] Clean up / split up NewStyle
- [x] Document NewStyle, remove old code and TODOs, and add type signatures
- [x] Remove old Runtime, Style, Shapes, Functions, Computation files and their tests
- [ ] Write actual tests for new Style, Optimizer, Shapes
- [x] Document ordering
- [ ] Revise how the compiler works in the design paper

# High-level steps
- [x] "Service" language grammar
- [x] Full Style parser
- [x] Selector statics 
- [x] Matching dynamics
- [x] Translation dynamics
- [ ] Block statics
- [ ] Selector specificity
- [x] Computational graph building, checking, and inference
- [ ] Tests for the new features
- [x] Better error messages similar to that of the DSLL
- [x] Substance - Style interface refactoring

# Implementation notes and TODOs
## Style parser
- TODOs
  - [x] Parsing selectors
  - [x] Parsing expressions
  - [x] Parsing blocks
  - [x] Parsing bindings
  - [x] Parsing lists
  - [x] Parsing Layer stmts
- Error handling can be done with `SourcePos` type returned by `getPosition` from Megaparsec. The question is how to do it. Do we want to change the AST types to records instead of sum types and add the positions as fields? Might be messy to write
- Permutation parsers are magical!

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran [Haddock](https://github.com/penrose/penrose/wiki/Writing-and-generating-documentation#generating-html-documentation-site) and there were no errors when generating the HTML site
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass locally using `stack test`
- [ ] My code follows the style guidelines of this project (e.g.: no HLint warnings)

### CODE REVIEW

**Comment by hypotext** [comment]
@wodeni I think this branch is ready to merge to master. Would you like to review it first? If not, that's no problem, I'll go ahead and merge.

Also, I documented the compiler [here](https://github.com/penrose/penrose/wiki/How-the-Style-compiler-works). I wrote or started some sections on the things you worked on (labels, shapes, and rendering) and tagged you there. Can you fill in those sections and/or check that they're correct?

**Comment by wodeni** [comment]
@hypotext I'm happy to review the changes before syncing tomorrow. Anything you want me to look at in particular (e.g.: the largest change since I left).

I will also take a look at the compiler doc.

p.s.: we should also fix Travis before merging. Let me know if you want to add any tests.

**Comment by hypotext** [comment]
@wodeni I reorganized NewStyle into Style and GenOptProblem, so take a look at GenOptProblem and let me know if the code organization there makes sense. I haven't changed ShapeDef (now called Shapes) so if you want to clean up code there, now's the time 🙂 

I'd like to add some tests eventually (but I won't have time this week) so it would be good to have Travis working again. Thanks!

My priority for the next few weeks is going to be on designing/implementing things for the new domains (real analysis and raytracing).

**Comment by hypotext** [comment]
I may make more edits to the wiki, so before you read it, just make sure you have the most recent version.

**Comment by wodeni** [comment]
@dormaayan it was not implemented. I do see it in the grammar though, weird. Will take a look

**Comment by wodeni** [comment]
Added the functionality in f009f5f. @dormaayan 

I tested the new parser with the following Style file:
```
Set u {
    u.shape = Circle { }
    u.shape.endX = u.startX + 2^6 + len(u.shape)
    u.shape.endX = len(1 + len(2))
    u.shape.endX = len(1 + len(2)) + 1
    u.shape.endX = len(v.shape) + u.shape.startX
    u.shape.endX = len(u.shape) + 1.0
}
```

Here is the AST generated:
```
Style AST:

[ ( Select
      Selector
        { selHead =
            [ PatternDecl' (STCtor "Set" []) (BStyVar (StyVar' "u")) ]
        , selWith = []
        , selWhere = []
        , selNamespace = Nothing
        }
  , [ Assign
        (FieldPath (BStyVar (StyVar' "u")) "shape") (Ctor "Circle" [])
    , Assign
        (PropertyPath (BStyVar (StyVar' "u")) "shape" "endX")
        (BinOp
           BPlus
           (BinOp
              BPlus
              (EPath (FieldPath (BStyVar (StyVar' "u")) "startX"))
              (BinOp Exp (IntLit 2) (IntLit 6)))
           (CompApp
              "len" [ EPath (FieldPath (BStyVar (StyVar' "u")) "shape") ]))
    , Assign
        (PropertyPath (BStyVar (StyVar' "u")) "shape" "endX")
        (CompApp
           "len" [ BinOp BPlus (IntLit 1) (CompApp "len" [ IntLit 2 ]) ])
    , Assign
        (PropertyPath (BStyVar (StyVar' "u")) "shape" "endX")
        (BinOp
           BPlus
           (CompApp
              "len" [ BinOp BPlus (IntLit 1) (CompApp "len" [ IntLit 2 ]) ])
           (IntLit 1))
    , Assign
        (PropertyPath (BStyVar (StyVar' "u")) "shape" "endX")
        (BinOp
           BPlus
           (CompApp
              "len" [ EPath (FieldPath (BStyVar (StyVar' "v")) "shape") ])
           (EPath (PropertyPath (BStyVar (StyVar' "u")) "shape" "startX")))
    , Assign
        (PropertyPath (BStyVar (StyVar' "u")) "shape" "endX")
        (BinOp
           BPlus
           (CompApp
              "len" [ EPath (FieldPath (BStyVar (StyVar' "u")) "shape") ])
           (AFloat (Fix 1.0)))
    ]
  )
]
```

File: src/GenOptProblem.hs [hunk 1]
```diff
@@ -0,0 +1,706 @@
...
+
+instance Eq OptStatus where
+         x == y = case (x, y) of
+                  (NewIter, NewIter) -> True
+                  (EPConverged, EPConverged) -> True
+                  (UnconstrainedRunning a, UnconstrainedRunning b) -> a == b
+                  (UnconstrainedConverged a, UnconstrainedConverged b) -> a == b
+                  (_, _) -> False
+
+data Params = Params { weight :: forall a . (Autofloat a) => a,
```
Comment (wodeni) [inline src/GenOptProblem.hs]: Do weights need to be changed to `[Float]` as well? We are okay for now because we don't really use them (?)

Comment (hypotext) [reply src/GenOptProblem.hs]: Done

**Pull request merged**